### PR TITLE
Add Motion Blur #302

### PR DESCRIPTION
**Pull request by mronta**

### CODE REVIEW

**Comment by mronta** [comment]
The tests are working on my local machine. Is there anything i am missing?

**Comment by edgarriba** [comment]
@mronta from which branch did you checkout ?

File: kornia/filters/motion.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        assert border_type in ["constant", "reflect", "replicate", "circular"]
+        self.border_type = border_type
+        self.kernel = torch.unsqueeze(get_motion_kernel2d(kernel_size, angle, direction), dim=0)
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__} (kernel_size={self.kernel_size}, ' \
+               f'angle={self.angle}, direction={self.direction})'
+
+    def forward(self, x: torch.Tensor):  # type: ignore
+        return filter2D(x, self.kernel, self.border_type)
```
Comment (edgarriba) [inline kornia/filters/motion.py]: @mronta be aware of this https://github.com/kornia/kornia/issues/321

File: test/filters/test_motion.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+from torch.testing import assert_allclose
+
+
+@pytest.mark.parametrize("ksize", [3, 11])
+@pytest.mark.parametrize("angle", [0., 360.])
+@pytest.mark.parametrize("direction", [-1., 1.])
+def test_get_motion_kernel2d(ksize, angle, direction):
+    kernel = kornia.get_motion_kernel2d((ksize, ksize), angle, direction)
+    assert kernel.shape == (ksize, ksize)
+    assert kernel.sum().item() == pytest.approx(1.0)
```
Comment (edgarriba) [inline test/filters/test_motion.py]: use `assert_allclose`

File: test/filters/test_motion.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+class TestMotionBlur:
+    @pytest.mark.parametrize("batch_shape", [(1, 4, 8, 15), (2, 3, 11, 7)])
+    @pytest.mark.parametrize("device_type", ("cpu", "cuda"))
+    def test_motion_blur(self, batch_shape, device_type):
```
Comment (edgarriba) [inline test/filters/test_motion.py]: don't use `device_type` yet, this is coming in a new PR https://github.com/kornia/kornia/pull/313

File: kornia/filters/kernels.py [hunk 4]
```diff
@@ -3,6 +3,8 @@
 import torch
 import torch.nn as nn
 
+from ..geometry import transform as transf
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/kernels.py]: Fixed

File: kornia/filters/kernels.py [hunk 5]
```diff
@@ -300,3 +302,25 @@ def get_laplacian_kernel2d(kernel_size: int) -> torch.Tensor:
     kernel[mid, mid] = 1 - kernel_size ** 2
     kernel_2d: torch.Tensor = kernel
     return kernel_2d
+
+
+def get_motion_kernel2d(
+        kernel_size: Tuple[int, int], angle: float, direction: float = 0
+) -> torch.Tensor:
+    if kernel_size[0] != kernel_size[1]:
+        raise ValueError("Kernel must be squared: height and width different")
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/kernels.py]: Fixed

File: kornia/filters/kernels.py [hunk 6]
```diff
@@ -300,3 +302,25 @@ def get_laplacian_kernel2d(kernel_size: int) -> torch.Tensor:
...
+    if kernel_size[0] % 2 == 0 or kernel_size[0] < 3:
+        raise ValueError("Kernel width and height must be odd and >= than 3")
+
+    def clip(value, min_value, max_value):
+        return max(min_value, min(value, max_value))
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/kernels.py]: It was an attempt to avoid the tensor conversion of 'direction', but for sure with clamp the code is more polished. Fixed

File: kornia/filters/kernels.py [hunk 7]
```diff
@@ -300,3 +302,25 @@ def get_laplacian_kernel2d(kernel_size: int) -> torch.Tensor:
...
+    # direction from [-1, 1] to [0, 1] range
+    direction = (clip(direction, -1., 1.) + 1.) / 2.
+    kernel = torch.zeros(kernel_size, dtype=torch.float)
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/kernels.py]: Yes the kernel is a 2d tensor. Are `device` and `dtype` needed? Other filters do not have these

File: kornia/filters/kernels.py [hunk 8]
```diff
@@ -300,3 +302,25 @@ def get_laplacian_kernel2d(kernel_size: int) -> torch.Tensor:
     kernel[mid, mid] = 1 - kernel_size ** 2
     kernel_2d: torch.Tensor = kernel
     return kernel_2d
+
+
+def get_motion_kernel2d(
+        kernel_size: Tuple[int, int], angle: float, direction: float = 0
+) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/kernels.py]: Done

File: kornia/filters/motion.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
+from typing import Tuple
+
+import torch
+import torch.nn as nn
+
+from ..filters.kernels import get_motion_kernel2d
+from ..filters.filter import filter2D
```
[the remark this answers is not in the document]
Comment (mronta) [reply kornia/filters/motion.py]: Fixed

File: test/filters/test_motion.py (hunk 2 above)
Comment (mronta) [reply test/filters/test_motion.py]: I made a fix. Not 100% sure if it was what you were meaning

File: test/filters/test_motion.py (hunk 3 above)
Comment (mronta) [reply test/filters/test_motion.py]: Fixed

File: kornia/filters/motion.py (hunk 1 above)
Comment (mronta) [reply kornia/filters/motion.py]: `torch.unsqueeze(kernel, dim=0)` is done before filter2D so this should not be a problem

File: kornia/filters/kernels.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/filters/kernels.py]: right. This something that can be improved/discussed later
